### PR TITLE
`Listener` refactor, `refresh` request handling

### PR DESCRIPTION
**Pull request by LVala**

### CODE REVIEW

File: lib/ex_turn/allocation_handler.ex [hunk 1]
```diff
@@ -96,12 +106,63 @@ defmodule ExTURN.AllocationHandler do
     {:noreply, state}
   end
 
+  @impl true
+  def handle_info(:check_expiration, state) do
+    if System.os_time(:second) >= state.expiry_timestamp do
+      Logger.info("Allocation expired")
```
Comment (mickel8) [inline lib/ex_turn/allocation_handler.ex]:
```suggestion
      Logger.info("Allocation expired. Shutting down allocation handler.")
```

File: lib/ex_turn/allocation_handler.ex [hunk 2]
```diff
@@ -96,12 +106,63 @@ defmodule ExTURN.AllocationHandler do
...
+      {:stop, {:shutdown, :allocation_expired}, state}
+    else
+      {:noreply, state}
+    end
+  end
+
   @impl true
   def handle_info(msg, state) do
-    Logger.warn("Got unexpected msg: #{inspect(msg)}")
+    Logger.warning("Got unexpected OTP message: #{inspect(msg)}")
```
Comment (mickel8) [inline lib/ex_turn/allocation_handler.ex:126]: Why OTP?

File: lib/ex_turn/allocation_handler.ex [hunk 3]
```diff
@@ -96,12 +106,63 @@ defmodule ExTURN.AllocationHandler do
...
+            {:noreply, state}
+          end
+
+        {ret_val, response}
+      else
+        {:error, response} -> {{:noreply, state}, response}
+      end
+
+    :gen_udp.send(state.turn_socket, c_ip, c_port, Message.encode(response))
+    ret_val
```
Comment (mickel8) [inline lib/ex_turn/allocation_handler.ex]: I would consider not to assign value returned from `with` but instead send response inside with but maybe your approach is less error-prone :thinking:  Any thoughts?

File: lib/ex_turn/utils.ex [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  def get_lifetime(msg) do
+    case Message.get_attribute(msg, Lifetime) do
+      {:ok, %Lifetime{lifetime: lifetime}} ->
+        desired_lifetime =
+          if lifetime == 0, do: 0, else: max(@default_lifetime, min(lifetime, @max_lifetime))
+
+        {:ok, desired_lifetime}
+
+      nil ->
+        {:ok, @default_lifetime}
```
Comment (mickel8) [inline lib/ex_turn/utils.ex:29]: That's a little bit unintuitive. Would changing get_lifetime(msg) into get_lifetime(msg, default) be better?

File: lib/ex_turn/auth.ex [hunk 5]
```diff
@@ -16,48 +16,52 @@ defmodule ExTURN.Auth do
...
 
-      {:error, :stale_nonce} ->
-        Logger.info("Stale nonce. Rejecting.")
-        {:error, build_error(msg.transaction_id, msg.type.method, 438, with_attrs?: true)}
+        {:error, :stale_nonce} ->
+          {"Stale nonce, rejected", 438, true}
+
+        :error ->
+          {"Bad message integrity, rejected", 401, true}
+      end
```
Comment (mickel8) [inline lib/ex_turn/auth.ex]: Very nice and clean approach! I wonder if moving `{:error, reason` -> `{msg, code, with_attrs?}` translation into build_error would make this even better?  This build_error would have to call Utils.build_error under the hood

File: lib/ex_turn/allocation_handler.ex [hunk 6]
```diff
@@ -186,11 +232,11 @@ defmodule ExTURN.AllocationHandler do
 
         :gen_udp.send(state.turn_socket, c_ip, c_port, response)
 
-        state
+        {:noreply, state}
```
Comment (mickel8) [inline lib/ex_turn/allocation_handler.ex]: That's a little bit questionable. Personally, I like to split GenServer logic from functions that process some data. This way it's easy to move those functions into another module or add some additional logic before returning from GenServer's callback. But I don't insist, just wanted to share my point of view

File: lib/ex_turn/allocation_handler.ex (hunk 2 above)
Comment (LVala) [reply lib/ex_turn/allocation_handler.ex:126]: I wanted to emphasise that this is not STUN/TURN message, but Erlang/Elixir/OTP message that the process received, not sure which word to use.

File: lib/ex_turn/allocation_handler.ex (hunk 6 above)
Comment (LVala) [reply lib/ex_turn/allocation_handler.ex]: Now that I think about it, I agree.

**Review by mickel8 — APPROVED** [review]

**Pull request merged**